### PR TITLE
Added trace and diag with batch support for linalg crate

### PR DESCRIPTION
**Pull request by niklund**
### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Relevant issue:
https://github.com/tracel-ai/burn/issues/1538

### Changes

The feature description of the issue is to add common linear algebra operations. There are some existing in
the linalg crate such as vector norms and cosine similarity. I  added trace and diag operation for starters. Will continue working on more operations. 
### Testing

Test cases covering rank 2, 3 and 4. Square tall and wide matrices and edge cases such as 1x1, single row/column with multiple data types. 
Tested with: cargo test -p burn-ndarray diag && cargo test -p burn-ndarray trace
cargo run-checks passes.

### CODE REVIEW

File: crates/burn-tensor/src/tensor/linalg/trace.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+/// # Arguments
+///
+/// * `tensor` - The input tensor with at least 2 dimensions.
+///
+/// # Returns
+///
+/// Tensor with same shape as the input, where the last two dimensions are reduced to size 1,
+/// containing the trace of each matrix.
+///
+pub fn trace<B: Backend, const D: usize>(tensor: Tensor<B, D>) -> Tensor<B, D> {
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/linalg/trace.rs]: Could this simply be implemented as a sum of tensor diag?

Comment (niklund) [reply crates/burn-tensor/src/tensor/linalg/trace.rs]: Yes for sure that was a big oversight not using diag for the trace. Makes total sense.

Regarding the rank preservation I implemented it like that since the vector_norms also preserved the rank so I figured that was the way to go, but I think it actually makes more sense to reduce the rank. I am completely open to any which way. Also 

> Also, we should assert that D >= 2.

Also agree with this!

Thank you for the feedback, I would appreciate input as to which way of handling the rank is preferable.

Comment (laggui) [reply crates/burn-tensor/src/tensor/linalg/trace.rs]: > Regarding the rank preservation I implemented it like that since the vector_norms also preserved the rank

That's a fair assumption I think. But that's only because it reduces a dim down to a scalar, and these reductions currently preserve rank.

For `diag`, the operation effectively "consumes" a dimension to return the diag vector so it makes sense to return `D - 1`. The trace is then a reduction over that, so this will match the current reduction pattern and the output will also be `D - 1` in this case.

Comment (niklund) [reply crates/burn-tensor/src/tensor/linalg/trace.rs]: Ah yes that makes total sense. I ended up doing it like that now. Thank you very much for your feedback!

File: crates/burn-tensor/src/tensor/linalg/diag.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+/// treating them as the matrix dimensions, while preserving all leading batch dimensions.
+///
+/// # Arguments
+///
+/// * `tensor` - The input tensor with at least 2 dimensions.
+///
+/// # Returns
+/// Tensor with rank D - 1, where the last two matrix dimensions are replaced by a single
+/// dimension containing the diagonal elements
+///
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/linalg/diag.rs]:
```suggestion
/// A tensor of rank `D - 1`, where the last dimension contains the diagonal elements of the input.
```

File: crates/burn-tensor/src/tensor/linalg/trace.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+///
+/// # Arguments
+///
+/// * `tensor` - The input tensor with at least 2 dimensions.
+///
+/// # Returns
+///
+/// Tensor with rank D - 1, where the last two matrix dimensions are replaced by a single
+/// dimension containing the the trace for each matrix.
+///
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/linalg/trace.rs]:
```suggestion
/// A tensor of rank `D - 1`, where the last dimension contains the sum along diagonals of the input.
```

File: crates/burn-tensor/src/tensor/linalg/trace.rs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+use super::diag;
+use crate::backend::Backend;
+use crate::tensor::Tensor;
+
+/// Computes the trace of the of square matrices.
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/linalg/trace.rs]: I think the "square" here is over specified no? You also have tests that work on non square matrices.

**Review by laggui — CHANGES_REQUESTED** [review]
Some minor comments on doc, otherwise LGTM!

**Review by laggui — APPROVED** [review]
